shared/lib_battery_voltage.cpp

```diff
@@ -342,13 +350,30 @@ void voltage_dynamic_t::parameter_compute() {
 }
 
 void voltage_dynamic_t::set_initial_SOC(double init_soc) {
-    updateVoltage(init_soc * 0.01 * params->dynamic.Qfull, params->dynamic.Qfull, 0, 25, params->dt_hr);
+    updateVoltage(init_soc * 0.01 * params->dynamic.Qfull * params->num_strings, params->dynamic.Qfull * params->num_strings, 0, 25, params->dt_hr);
 }
 
 // everything in here is on a per-cell basis
 double voltage_dynamic_t::voltage_model_tremblay_hybrid(double Q_cell, double I, double q0_cell) {
     double it = Q_cell - q0_cell;
-    double E = _E0 - _K * (Q_cell / (Q_cell - it)) + _A * exp(-_B0 * it);
+    //double it = params->dynamic.Qfull_mod - q0_cell;
+    double Q_cell_mod, C, x;
+    if (params->dynamic.Vcut != 0) {
+        C = (-1 * params->dynamic.Vcut + _E0 - params->resistance * I + _A * exp(-_B0 * Q_cell)) / _K;
+        x = Q_cell / (C - 1);
+        Q_cell_mod = Q_cell + x;
+        
+    }
+    else {
+        Q_cell_mod = Q_cell;
+        params->dynamic.Qfull_mod = 0;
+        
+    }
+    //->dynamic.Qfull_mod = Q_cell_mod;
+    //double E = _E0 - _K * (Q_cell / (Q_cell - it)) + _A * exp(-_B0 * it);
+    
+    double E = _E0 - _K * (Q_cell_mod / (Q_cell_mod - it)) + _A * exp(-_B0 * it);
+    //double E = _E0 - _K * (params->dynamic.Qfull_mod / (params->dynamic.Qfull_mod - it)) + _A * exp(-_B0 * it);
```

Comment:
@brtietz @dguittet Ended up putting the modification to Qfull here to account for the potential change in capacity due to the lifetime and thermal models. If Qfull_mod is calculated in parameter_compute and not adjusted for changes in maximum battery capacity, the maximum possible discharge will be greater than what is actual possible. Ran into this issue when existing tests with 0 cutoff voltage were failing due to a difference between the qmax that was originally passed and the non-changing Qfull_mod. Does this track?

Reply:
Yes, Q_cell_mod should be adjusted for thermal and lifetime changes, which means it should be recomputed on each call to this function.

Reply:
Looks right to recalculate it per call given how Q_cell changes. I think that mean the `Q_full_mod` should not be within the `params` struct, since those are values that are consistent throughout the simulation. It should be within the `state` struct instead

Reply:
However, `voltage_model_tremblay_hybrid` should not update the `state` since this function is used in many "prospective" functions such as all the calculate_x functions. These prospective functions are used to query what is a possible next state of the battery without actually moving the battery state to any of them. The only function that can "step forward" the battery is `updateVoltage`.

```
    double Vfull = 4.1;
    double Vexp = 4.05;
    double Vnom = 3.4;
    double Qfull = 2.25;
    double Qexp = 0.04;
    double Qnom = 2.0;
    double C_rate = 0.2;
```

This gives us 3 points on the V vs Q curve: (V_full, 0), (V_exp, Q_full - Q_exp), (V_nom, Q_full - Q_nom).

Now, however, we have 4 different points: (V_full, 0), (V_exp, Q_full - Q_exp), (V_nom, Q_full - Q_nom), (V_cutoff, Q_full)

Once the parameters have been generated, we calculate an additional "parameter" Q_full_mod.

When the Q_full is decreased due to thermal or lifetime, the point (V_cutoff, Q_full) may no longer hold as we'll reach the same V_cutoff but with less Q_full. And so `Q_full_mod` needs to be updated whenever the battery state is changed.

This update of Q_full and Q_full_mod can occur just in `updateVoltage`, before calling `voltage_model_tremblay_hybrid`. The latter should take Q_full_mod as given. All the other calculate_x functions should do the same.

For `calculate_max_discharge_w` the reason we have to do the while loop was because the highest current (I = q) doesn't necessarily produce the highest discharge power. So the loop iterated from I = 0 to q0 * dt, and I think that's still the same now since q0 <= Q_full < Q_full_mod, but the `voltage_model_tremblay_hybrid` will be using Q_full_mod.

For `calculate_current_for_target_w` the `solve_current_for_discharge_power` and `solve_current_for_charge_power` may need some way to enforce the V_cutoff if it's not matching up with `calculate_max_discharge_w`. I'll have to think a bit more later.